src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -295,47 +310,106 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
         {
             FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructor, out int?[] parameterMap);
 
+#if NET8_0_OR_GREATER
+            ConstructorInvoker invoker = ConstructorInvoker.Create(constructor);
+#endif
+
             ParameterInfo[] constructorParameters = constructor.GetParameters();
             if (constructorParameters.Length == 0)
             {
                 return (IServiceProvider serviceProvider, object?[]? arguments) =>
+#if NET8_0_OR_GREATER
+                    invoker.Invoke()!;
+#else
                     constructor.Invoke(BindingFlags.DoNotWrapExceptions, binder: null, parameters: null, culture: null);
+#endif
             }
 
-            FactoryParameterContext[] parameters = new FactoryParameterContext[constructorParameters.Length];
+            Type declaringType = constructor.DeclaringType!;
+
+#if NET8_0_OR_GREATER
+            bool useFixedValues = constructorParameters.Length <= FixedArgumentThreshold;
+            bool hasAnyDefaultValues = false;
+            int matchedArgCount = 0;
+            int matchedArgCountWithMap = 0;
```

Comment:
It might make for easier reading code if you refactored each side of this `#if` condition into self-contained methods. This is a lot to take in for a single method.

Maybe it also might make sense to move this all to the ".Reflection.cs" file, since all this uses reflection.

Reply:
I kept one method due to shared code, but combined the >v8 code to make that easier to understand.